dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/ApacheHttpClientBlockingChannel.java

```diff
@@ -473,6 +475,42 @@ public String toString() {
         public void close() {}
     }
 
+    /**
+     * {@link ModulatingOutputStream} limits the size of individual writes to the wrapped {@link OutputStream}
+     * in order to prevent degraded performance on large buffers as described in
+     * <a href="https://github.com/palantir/hadoop-crypto/pull/586">hadoop-crypto#586</a>.
+     */
+    static final class ModulatingOutputStream extends FilterOutputStream {
+
+        /**
+         * Block size of 16 KB is small enough to allow cipher implementations to become hot and optimize properly
+         * when given large inputs. Otherwise large array writes into a {@link javax.crypto.CipherOutputStream} fail to
+         * use intrinsified implementations. If 16 KB blocks aren't enough to produce hot methods, the I/O is small
+         * and infrequent enough that performance isn't relevant.
+         * For more information, see the details around {@code com.sun.crypto.provider.GHASH::processBlocks} in
+         * <a href="https://github.com/palantir/hadoop-crypto/pull/586#issuecomment-964394587">
+         * hadoop-crypto#586 (comment)</a>
+         */
+        private static final int BLOCK_SIZE = 16 * 1024;
+
+        ModulatingOutputStream(OutputStream delegate) {
+            super(delegate);
+        }
+
+        @Override
+        public void write(byte[] buffer, int off, int len) throws IOException {
+            Objects.checkFromIndexSize(off, len, buffer.length);
+            int currentOffset = off;
+            int remaining = len;
+            while (remaining > 0) {
+                int toWrite = Math.min(remaining, BLOCK_SIZE);
+                out.write(buffer, currentOffset, toWrite);
+                currentOffset += toWrite;
+                remaining -= toWrite;
+            }
+        }
+    }
```

Comment:
Might be helpful to override `write(int)` as a simple passthrough to `out` and avoid array allocation

Reply:
Ah right, sorry for the runaround!